rp2040-hal/src/pwm.rs

```diff
@@ -1,260 +1,324 @@
+}
 
-        if (pin_num == 255) {
-            pin_num = $pins[0];
-        }
+/// The `Pwm` type respresents an abstraction of the PWM hardware block.
+pub struct Pwm {
+    regs: pac::PWM,
+}
 
-        Self {
-            pin: pin_num,
-        }
+impl Pwm {
+    /// Generate a new `Pwm` instance which takes ownersip of the provided `rp2040_pac::PWM`
+    /// object.
+    pub fn new(pac_pwm: pac::PWM, resets: &mut pac::RESETS) -> Self {
+        pac_pwm.reset_bring_up(resets);
+        Pwm { regs: pac_pwm }
     }
 
-    // TODO: This function contains all the methods that required the PWM to have an instance of PADS_BANK0, RESETS, or IO_BANK0.
-    // Since the GPIO pins take ownership of these, after the GPIO refactor, this method should be moved into gpio.rs, and the PWM
-    // will instead receive a single gpio pin.
-    fn init_io(&self, pwm: &mut pac::$PWMX, pad : &mut pac::PADS_BANK0, io : &mut pac::IO_BANK0, resets: &mut pac::RESETS) -> () {
-        //TODO: Merge these into gpio.rs split function after GPIO refactor. At the moment, this is here because these need to be reset for
-        // the PWM to work. However, because they're here, they'll be reset every time a new PWM pin is created (BAD).
-        pwm.reset_bring_up(resets);
-        io.reset_bring_up(resets);
-
-        pad.gpio[self.pin].write(|w| w.ie().set_bit());
-        pad.gpio[self.pin].write(|w| w.od().clear_bit());
-        unsafe {
-            io.gpio[self.pin].gpio_ctrl.write_with_zero(|w| w.funcsel().pwm_a_0());
+    /// Generate a `PwmSlice` instance if the provided slice index is valid. Note that the index is
+    /// 0 based.
+    pub fn slice(&self, idx: usize) -> Option<PwmSlice> {
+        if idx > 7 {
+            // The rp2040 PWM block has only 8 slices indexed from 0 to 7
+            return None;
         }
+
+        Some(PwmSlice::new(self, idx))
     }
+}
 
-    fn cc(&self) -> &pac::$pwmx::ch::CC {
-        unsafe {
-            &(*pac::$PWMX::ptr()).ch[$i].cc
-        }
+/// The `PwmSlice` type represents an abstraction of a PWM slice of the rp2040 PWM block.
+pub struct PwmSlice<'a> {
+    pwm: &'a Pwm,
+    idx: usize,
+}
+
+impl<'a> PwmSlice<'a> {
+    fn new(pwm: &'a Pwm, idx: usize) -> Self {
+        PwmSlice { pwm: pwm, idx: idx }
     }
 
-    fn csr(&self) -> &pac::$pwmx::ch::CSR {
-        unsafe {
-            &(*pac::$PWMX::ptr()).ch[$i].csr
-        }
+    /// Disable the phase correct mode of the PWM slice.
+    #[inline(always)]
+    pub fn disable_phase_correct(&self) {
+        self.pwm.regs.ch[self.idx]
+            .csr
+            .write(|w| w.ph_correct().clear_bit());
     }
 
-    fn ctr(&self) -> &pac::$pwmx::ch::CTR {
-        unsafe {
-            &(*pac::$PWMX::ptr()).ch[$i].ctr
-        }
+    /// Enable the phase correct mode of the PWM slice.
+    #[inline(always)]
+    pub fn enable_phase_correct(&self) {
+        self.pwm.regs.ch[self.idx]
+            .csr
+            .write(|w| w.ph_correct().set_bit());
```

Comment:
`write` is not what you use for setting individual bits in a register if you want to preserve the rest of the register fields.
```rust
self.pwm.regs.ch[self.idx].csr.write(|w| w.ph_correct().set_bit());
```
Inside the closure, w starts with the reset value of the register, and then you set the ph_correct bit, and that gets written back to the register (clearing all the other set bits of the register).
You want to use `modify` in cases such as this:
```rust
self.pwm.regs.ch[self.idx].csr.modify(|_,w| w.ph_correct().set_bit());
```
Note how we don't care about the value of r in this case, so it is replaced with `_` in the capture clause and is inaccessible inside the closure body. This lets others  know that it is unused.
You can read more here about this here: https://docs.rust-embedded.org/book/start/registers.html#modifying